libyul/Object.cpp

```diff
@@ -61,13 +62,49 @@ string Object::toString(Dialect const* _dialect) const
 	return "object \"" + name.str() + "\" {\n" + indent(inner) + "\n}";
 }
 
-set<YulString> Object::dataNames() const
+set<YulString> Object::qualifiedDataNames() const
 {
-	set<YulString> names;
-	names.insert(name);
-	for (auto const& subObject: subIndexByName)
-		names.insert(subObject.first);
-	// The empty name is not valid
-	names.erase(YulString{});
-	return names;
+	set<YulString> qualifiedNames = {name};
+	for (shared_ptr<ObjectNode> const& subObjectNode: subObjects)
+	{
+		yulAssert(qualifiedNames.count(subObjectNode->name) == 0, "");
+		qualifiedNames.insert(subObjectNode->name);
+		if (auto const* subObject = dynamic_cast<Object const*>(subObjectNode.get()))
+			for (YulString const& subSubObj: subObject->qualifiedDataNames())
+				if (subObject->name != subSubObj)
+				{
+					yulAssert(qualifiedNames.count(YulString{subObject->name.str() + "." + subSubObj.str()}) == 0, "");
+					qualifiedNames.insert(YulString{subObject->name.str() + "." + subSubObj.str()});
+				}
+	}
+
+	qualifiedNames.erase(YulString{});
+	return qualifiedNames;
+}
+
+vector<size_t> Object::pathToSubObject(string _qualifiedName) const
+{
+	yulAssert(_qualifiedName != name.str(), "");
+
+	if (boost::algorithm::starts_with(_qualifiedName, name.str() + "."))
+		_qualifiedName = _qualifiedName.substr(name.str().length() + 1);
+
+	vector<string> subObjectPathStr;
+	boost::algorithm::split(subObjectPathStr, _qualifiedName, boost::is_any_of("."));
+
+	vector<size_t> path;
+	Object const* object = this;
+	for (string const& currentSubObjectName: subObjectPathStr)
+	{
+		auto subIndexIt = object->subIndexByName.find(YulString{currentSubObjectName});
+		yulAssert(
+			subIndexIt != object->subIndexByName.end(),
+			"Could not find assembly object <" + _qualifiedName + ">."
+		);
+		object = dynamic_cast<Object const*>(object->subObjects[subIndexIt->second].get());
+		yulAssert(object, "Could not find assembly object <" + _qualifiedName + ">.");
+		path.push_back({object->subId});
```

Comment:
`subObjects` contain data objects as well.
So, if you have in subObjects: data object1 object2
index in subObjects for object1 and object2 would be 1 and 2, but subId in Assembly is 0 and 1 as data is treated differently...

This is why I added data before subobjects in my test(test/libyul/objectCompiler/subObjectAccess.yul). Without subId, it was failing that test.

Reply:
Ok, that makes sense. But this ID still seems to me like something that could easily go out of sync with the info stored in the parent object. Maybe adding a `parent` field and finding the ID in the map by `name` would be better?

Anyway, if `subId` is there to stay, I'd actually add a few asserts (here and maybe in some other places where the parent object is available) that `subId <= subObjects.size()`.

Reply:
I think it would create dependency from object to Assembly if I add parent, which is not perfect.
I do not think it can go out of sync, as when we compile Assemblies we assign sub object's subId

```
for (auto const& subNode: _object.subObjects)
	if (auto* subObject = dynamic_cast<Object*>(subNode.get()))
	{
		auto subAssemblyAndID = m_assembly.createSubAssembly();
		context.subIDs[subObject->name] = subAssemblyAndID.second;
		subObject->subId = subAssemblyAndID.second;
		compile(*subObject, *subAssemblyAndID.first, m_dialect, m_evm15, _optimize);
	}
        ...
```

Reply:
Why a dependency to `Assembly`? `parent` would be just `Object const*`. And the dependency already exists - `subId` depends on the content of parent's `subObjects` - it's just implicit.

I did not mean that it would go out of sync in the current code. Just that it would be easy to make a mistake and use an ID that does not really match the content of `subObjects` and there's not way to assert against that. But well, maybe tests would catch that too.

Anyway, it was just a suggestion. `parent` is not a perfect solution either so maybe let's leave it as is.